proposals/NNNN-linalg-matrix.md

```diff
@@ -0,0 +1,110 @@
+Subgroup operations. Related terms like Quad or derivatives have similar
+meaning in different scoping contexts. Vulkan has also recently introduced the
+term "cooperative" when talking about operations that require participation from
+multiple threads, these can be viewed much like derivative operations but across
+the full SIMD unit instead of a subset of threads.
+
+All of these terms refer to the way the underlying instructions execute, not
+necessarily what they do. One big part of this proposal is to take 5 steps back
+and talk about what they do: linear algebra.
+
+## Motivation
+
+HLSL has a Vulkan extension for SIMD matrix types [0021 - vk::Cooperative
+Matrix](0021-vk-coop-matrix.md), and DirectX had previewed a similar feature in
+SM 6.8 called [Wave Matrix](https://github.com/microsoft/hlsl-specs/pull/61).
+This proposal is aimed at merging the two into a unified language feature that
+can be supported on all platforms (with some platform-specific limitations).
+
+## Proposed solution
+
+Below is a proposed Pseudo-HLSL API. The proposal uses C++20 concepts to
+represent template type constraints so as to avoid needing SFINAE complications.
+
+```c++
+namespace hlsl {
+
+template <class T>
+concept ArithmeticScalar = std::is_arithmetic<T>::value;
+
+namespace linalg {
+
+template <typename ComponentTy, uint M, uint N>
+  requires ArithmeticScalar<ComponentTy>
+class Matrix {
+  template <typename NewCompTy> Matrix<NewCompTy, M, N> cast();
+
+  Matrix operator+(Matrix);
+  Matrix operator-(Matrix);
+  Matrix operator*(Matrix);
+  Matrix operator/(Matrix);
+
+  template <typename T>
+    requires ArithmeticScalar<T>
+  Matrix operator+(T);
+  template <typename T>
+    requires ArithmeticScalar<T>
+  Matrix operator-(T);
+  template <typename T>
+    requires ArithmeticScalar<T>
+  Matrix operator*(T);
+  template <typename T>
+    requires ArithmeticScalar<T>
+  Matrix operator/(T);
+
+  static Matrix Splat(ElTy Val);
+  static Matrix Load(ByteAddressBuffer Res, uint StartOffset, uint Stride,
+                     bool ColMajor, uint Align = sizeof(ComponentTy));
+  static Matrix Load(RWByteAddressBuffer Res, uint StartOffset, uint Stride,
+                     bool ColMajor, uint Align = sizeof(ComponentTy));
+
+  static Matrix Load(groupshared ElTy Arr[], uint StartIdx, uint Stride,
+                     bool ColMajor);
+
+  void Store(RWByteAddressBuffer Res, uint StartOffset, uint Stride,
+             bool ColMajor, uint Align = sizeof(ComponentTy));
+
+  void Store(groupshared ElTy Arr[], uint StartIdx, uint Stride, bool ColMajor);
+
+  void MultiplyAccumulate(const ref Matrix<T, N, M>);
+  void SumAccumulate(const ref Matrix<T, N, M>);
+};
+
+template <typename T, uint M, uint N, uint K>
+Matrix<T, M, N> Multiply(const ref Matrix<T, M, K>, const ref Matrix<T, K, N>);
+
+} // namespace linalg
+} // namespace hlsl
+```
+
+## Detailed design
+
+## Outstanding Questions

```

Comment:
* Support for packed types?
* Support for other number formats that aren't natively supported by HLSL?